### PR TITLE
Add batch get, update, delete methods to Storage and Blob

### PR DESCRIPTION
**Pull request by mziccard**
This PR follows the discussion in #219.
It adds `get`, `update` and `delete` methods that simplify doing batch gets/updates/deletes. Methods are added to both `Storage` and `Blob`.
Both unit and integration tests are added.

### CODE REVIEW

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 1]
```diff
@@ -256,4 +261,88 @@ public URL signUrl(long expirationTimeInSeconds, SignUrlOption... options) {
   public Storage storage() {
     return storage;
   }
+
+  /**
+   * Gets the requested blobs. If {@code infos.length == 0} an empty list is returned. If
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: To me the empty case sounds obvious (and I don't think we stated it in other places) but fine as is.

Actually, what I think is a problem is that in some cases we are returning immutable list (for empty or singleton)
and in others we don't.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java [hunk 2]
```diff
@@ -577,6 +577,68 @@ public URL signUrl(BlobInfo blobInfo, long expiration, SignUrlOption... options)
     }
   }
 
+  @Override
+  public List<BlobInfo> get(BlobInfo blobInfo1, BlobInfo blobInfo2, BlobInfo... blobInfos) {
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Wasn't `get(BlobInfo blobInfo, BlobInfo blobInfos)` not enough?

I assume that when called with get(blobInfo) the singular version would be picked vs the the vargs, no?

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java [hunk 3]
```diff
@@ -577,6 +577,68 @@ public URL signUrl(BlobInfo blobInfo, long expiration, SignUrlOption... options)
...
+    List<Tuple<StorageObject, Map<StorageRpc.Option, ?>>> toGet =
+        Lists.newArrayListWithCapacity(blobInfos.length + 2);
+    toGet.add(Tuple.<StorageObject, Map<StorageRpc.Option, ?>>of(blobInfo1.toPb(), optionMap()));
+    toGet.add(Tuple.<StorageObject, Map<StorageRpc.Option, ?>>of(blobInfo2.toPb(), optionMap()));
+    for (BlobInfo blobInfo : blobInfos) {
+      toGet.add(Tuple.<StorageObject, Map<StorageRpc.Option, ?>>of(blobInfo.toPb(), optionMap()));
+    }
+    StorageRpc.BatchResponse response =
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Though this is slightly more efficient I am not sure I favor it over using Batch on the service layer.

File: gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java [hunk 4]
```diff
@@ -106,9 +108,12 @@ public Y y() {
     public BatchRequest(Iterable<Tuple<StorageObject, Map<Option, ?>>> toDelete,
         Iterable<Tuple<StorageObject, Map<Option, ?>>> toUpdate,
         Iterable<Tuple<StorageObject, Map<Option, ?>>> toGet) {
-      this.toDelete = ImmutableList.copyOf(toDelete);
-      this.toUpdate = ImmutableList.copyOf(toUpdate);
-      this.toGet = ImmutableList.copyOf(toGet);
+      this.toDelete = ImmutableList.copyOf(
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java]: See my comment bellow regarding using the Batch on the service layer rather the RPC layer.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 5]
```diff
@@ -256,4 +261,88 @@ public URL signUrl(long expirationTimeInSeconds, SignUrlOption... options) {
...
+    }
+  }
+
+  /**
+   * Updates the requested blobs. If {@code infos.length == 0} an empty list is returned. If
+   * {@code infos.length > 1} a batch request is used to update blobs.
+   *
+   * @param storage the storage service used to issue the request
+   * @param infos the blobs to update
+   * @return a list of {@code Blob} objects. If a blob does not exist the corresponding item in the
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Here and in other places I think we should add the "access denied" case to the returning `null` case.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java (hunk 1 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: You are right. We'll have to use `Collections.unmodifiableList` as guava's `ImmutableList` does not support `null` as a value, right?

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Yes.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 6]
```diff
@@ -256,4 +261,88 @@ public URL signUrl(long expirationTimeInSeconds, SignUrlOption... options) {
...
+   * list is {@code null}.
+   */
+  public static List<Blob> update(final Storage storage, BlobInfo... infos) {
+    checkNotNull(storage);
+    checkNotNull(infos);
+    int length = infos.length;
+    switch (length) {
+      case 0:
+        return Collections.emptyList();
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Fine as is, but I wonder if that is useful as oppose to having the signature:
`update(final Storage storage, BlobInfo blobInfo, BlobInfo... infos) `

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java [hunk 7]
```diff
@@ -159,4 +165,129 @@ public void testSignUrl() throws Exception {
     replay(storage);
     assertEquals(url, blob.signUrl(100));
   }
+
+  @Test
+  public void testGetNone() throws Exception {
+    replay(storage);
+    assertTrue(Blob.get(storage).isEmpty());
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java]: See my comment above about method signature.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java [hunk 8]
```diff
@@ -159,4 +165,129 @@ public void testSignUrl() throws Exception {
...
+    List<Blob> result = Blob.get(storage, BLOB_INFO_ARRAY);
+    assertEquals(blobInfoList.size(), result.size());
+    for (int i = 0; i < blobInfoList.size(); i++) {
+      assertEquals(blobInfoList.get(i), result.get(i).info());
+    }
+  }
+
+  @Test
+  public void testGetSomeNull() throws Exception {
+    List<BlobInfo> blobInfoList = Arrays.asList(BLOB_INFO_ARRAY[0], null, BLOB_INFO_ARRAY[2]);
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java]: Do you think we should support that? null value is not allowed in the singular case and I wonder if we want to allow it
in the vargarg case (and potentially hide user errors...)

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 2 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: This was just to have the same signature as `update`.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 3 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: I have no strong opinion. I thought a bit on what to do here, still in doubt. For big collection this is surely more efficient though.

**Comment by aozarov** [comment]
Do you think we should also add similar methods to Bucket (e.g. for Bucket.update)?

Now, I noticed that we have `Bucket.getAll(String...)` and I wonder if signature should be changed to `Bucket.get(String, String...)`

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 3 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Yes, but my vote would favor simplicity. I think for big collections a better contribution would be to allow batch chunking.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java (hunk 8 above)
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java]: This does not test the case in which null is passed as an argument. It tests the case in which the underlying `storage.get(BlobInfo...)` return `null` values in the list (for not found, etc.)

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 2 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Sorry, I was proposing to change all the cases of method(one, two, rest) to method(one, rest). 
Do you see a problem with that?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java (hunk 8 above)
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/BlobTest.java]: Ah, sorry my bad.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 3 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Do you think the code would be simpler? Still need to build `BatchRequest` and convert from `Iterable<Result>`.

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Yes, but we have only one place when we deal with RPC layer batching.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 2 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Yes I think that we would have a problem with `update`. With the signatures:
```java
update(BlobInfo... , Options...)
update(BlobInfo, BlobInfo...)
```
The call `update(blobInfo)` is ambiguous.

**Comment by mziccard** [comment]
The methods in `Blob` have more sense as they provide a nicer interface to a useful mechanism (batch requests) to enforce its usage. Adding the methods to `Bucket` is only syntactic sugar (and I wonder even how much it could be used). I don't know, maybe we should add them anyway.

Yes we should probably change `Bucket.getAll(String...)` to `Bucket.get(String, String...)`.

Should we do all of these in this PR?

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 2 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Ah, I see. could be fixed by adding `update(BlobInfo)` but I guess also fine as is.

**Comment by aozarov** [comment]
Technically we could also add bucket modification/get to the batch but I agree a need for such feature is low.  Lets leave it for now (can be added later if requested).

Doing `Bucket.get(String, String...)` in this PR seems fine to me.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java (hunk 6 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: The problem of a signature like:
```
update(final Storage storage, BlobInfo blobInfo, BlobInfo... infos)
```
Is when the user has a list of `BlobInfo`. I think this is a quite common case and in general I would avoid splitting varargs. (We already had to do that in `Storage` and I still am not so convinced)

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Right, though if we think user is typically going to have collections (rather than just using
the varargs directly) then maybe we should add or replace with `Iterable<BlobInfo>`?

I am fine as is, and if we change would rather have it consistent.

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Let's keep it as is but I would consider adding support for `Iterable<BlobInfo>`.

File: gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java (hunk 4 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java]: Is this change still necessary?

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java]: No it's not. But it might be a reasonable change anyway.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java (hunk 2 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Did you consider adding `method(BlobInfo)` to remove ambiguity and supporting `method(BlobInfo...)` to allow easy iteration conversion (similar to what you did for `Blob`)?

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java]: Forgot:) See last commit (f1c6b04)

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java [hunk 9]
```diff
@@ -638,4 +646,32 @@ public static Builder builder() {
    * @see <a href="https://cloud.google.com/storage/docs/access-control#Signed-URLs">Signed-URLs</a>
    */
   URL signUrl(BlobInfo blobInfo, long expirationTimeInSeconds, SignUrlOption... options);
+
+  /**
+   * Gets the requested blobs. A batch request is used to perform this call.
+   *
+   * @param blobInfos blobs to get
+   * @return an immutable list of {@code BlobInfo} objects. If a blob does not exist or access to it
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java]: these methods could still throw StorageException if applying the batch failed.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 10]
```diff
@@ -256,4 +261,87 @@ public URL signUrl(long expirationTimeInSeconds, SignUrlOption... options) {
...
   }
+
+  /**
+   * Gets the requested blobs. If {@code infos.length == 0} an empty list is returned. If
+   * {@code infos.length > 1} a batch request is used to fetch blobs.
+   *
+   * @param storage the storage service used to issue the request
+   * @param infos the blobs to get
+   * @return an immutable list of {@code Blob} objects. If a blob does not exist or access to it has
+   * been denied the corresponding item in the list is {@code null}.
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: see my comment in Storage regarding throws StorageException

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 11]
```diff
@@ -256,4 +261,87 @@ public URL signUrl(long expirationTimeInSeconds, SignUrlOption... options) {
...
+  public static List<Blob> get(final Storage storage, BlobInfo... infos) {
+    checkNotNull(storage);
+    checkNotNull(infos);
+    int length = infos.length;
+    switch (length) {
+      case 0:
+        return Collections.emptyList();
+      case 1:
+        return Collections.singletonList(
+            new Blob(storage, storage.get(infos[0].bucket(), infos[0].name())));
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: I like this optimization but I think it is in-consistent with the list (as I think a single get returns null for NOT_FOUND, but
does throw the exception for ACL). A change in either direction is fine with me, but maybe a more desirable one is to leave the single get as is (so, one has a way to get a detail on a failure if desired by using that method).

**Comment by aozarov** [comment]
I like it now (with the service-level batching and consistent, clean vararg) much better.

I commented about 2 issues regarding exceptions

I actually now re-thinking about the `BlobId` and wonder if it is actually worth adding it (and use it when appropriate instead of `BlobInfo`, e.g. in delete[s], get[s],...)  provided we *also* provide the convenient methods (such as `get(bucket, name)`, BlobInfo.builder(bucket, name),...) .
What do you think (in any case, clearly not for this PR)?

**Comment by mziccard** [comment]
I fixed last isses.

As I said in #62 I think that `BlobId` will not provide major benefits to users. However, it might help to have more uniform APIs. I'll work on adding it tomorrow, I'll take care of keeping al convenience methods as:
- `storage.get(bucket, name)`
- `BlobInfo.builder(bucket, name)` (while dropping `BlobInfo.of(bucket, name)` in favor of `BlobId.of(bucket, name)`
- `storage.delete(bucket, name)`
- `storage.reader(bucket, name)`

I would also add an `id()` getter to `BlobInfo` without delete `name()` and `bucket()` so that to avoid writing `info.id().name()`.

**Comment by aozarov** [comment]
Great, PR looks good and I think we are on the same page regarding #62.
